graphstorm-processing/graphstorm_processing/data_transformations/dist_transformations/dist_hard_negative_transformation.py

```diff
@@ -0,0 +1,123 @@
+    ORDER_INDEX,
+    EXPLODE_HARD_NEGATIVE_VALUE,
+)
+
+from .base_dist_transformation import DistributedTransformation
+
+
+class DistHardEdgeNegativeTransformation(DistributedTransformation):
+    """Transformation to apply hard negative transformation.
+
+    Parameters
+    ----------
+    cols : Sequence[str]
+        List of column names to apply hard negative transformation to.
+    spark: SparkSession
+        The spark session.
+    hard_node_mapping_dict: dict
+        The mapping dictionary contain mapping file directory and edge type.
+        {
+            "edge_type": str
+                Edge type to apply hard negative transformation.
+            "mapping_path": str
+                Path to the raw node mapping.
+            "format_name": str
+                Parquet.
+        }
+    separator: str, optional
+        The separator for string input value. Only required when input value type is string.
+    """
+
+    def __init__(
+        self,
+        cols: Sequence[str],
+        spark: SparkSession,
+        hard_node_mapping_dict: dict,
+        separator: str = "",
+    ) -> None:
+        super().__init__(cols, spark)
+        self.cols = cols
+        assert len(self.cols) == 1, "Hard Negative Transformation only supports single column"
+        self.separator = separator
+        self.hard_node_mapping_dict = hard_node_mapping_dict
+        assert self.hard_node_mapping_dict, "edge mapping dict cannot be None for hard negative "
+
+    def apply(self, input_df: DataFrame) -> DataFrame:
+        assert self.spark
+        input_col = self.cols[0]
+        column_type = input_df.schema[input_col].dataType
+        if isinstance(column_type, StringType):
+            transformed_df = input_df.withColumn(input_col, split(col(input_col), self.separator))
+        else:
+            transformed_df = input_df
+        # Edge type should be (src_ntype:relation_type:dst_ntype)
+        # Only support hard negative for destination nodes. Get the node type of destination nodes.
+        # TODO: support hard negative for source nodes.
+        _, _, dst_type = self.hard_node_mapping_dict["edge_type"].split(":")
+        mapping_prefix = self.hard_node_mapping_dict["mapping_path"]
+        format_name = self.hard_node_mapping_dict["format_name"]
+        hard_negative_node_mapping = self.spark.read.parquet(
+            f"{mapping_prefix}{dst_type}/{format_name}/"
+        )
+        max_size = (
+            transformed_df.select(F.size(F.col(input_col)).alias(f"{input_col}_size"))
+            .agg(F.max(f"{input_col}_size"))
+            .collect()[0][0]
+        )
+
+        # TODO: Use panda series to possibly improve the efficiency
+        # Explode the original list and join node id mapping dataframe
+        transformed_df = transformed_df.withColumn(ORDER_INDEX, F.monotonically_increasing_id())
+        transformed_df = transformed_df.withColumn(
+            EXPLODE_HARD_NEGATIVE_VALUE, F.explode(F.col(input_col))
```

Comment:
Could results in extremely large DFs, let's add a comment here. The expected row count will be `num_nodes*avg(len_of_negatives)`